File: libs/ZmqLib/go/pub/publisher.go
package publisher

import (
	"io/ioutil"
	"time"

	"github.com/golang/glog"
	zmq "github.com/pebbe/zmq4"
)

// PublishList struct is used to initialize topic, metadata & frame
type PublishList struct {
	Topic     string
	MetaData  map[string]string
	FrameBlob string
}

// loadCertificates is used to load all required certs
func loadCertificates(publicKeysDir string, privateKeysDir string) (string, string, string, string) {
	clientPublicKey, err := ioutil.ReadFile(publicKeysDir + "/client.key")
	if err != nil {
		glog.Errorf("Error reading clientPublicKey : %v", err)
		return "", "", "", ""
	}
	clientPublic := string(clientPublicKey)

	clientSecretKey, err := ioutil.ReadFile(privateKeysDir + "/client.key_secret")
	if err != nil {
		glog.Errorf("Error reading clientSecretKey : %v", err)
		return "", "", "", ""
	}
	clientSecret := string(clientSecretKey)

	serverPublicKey, err := ioutil.ReadFile(publicKeysDir + "/server.key")
	if err != nil {
		glog.Errorf("Error reading serverPublicKey : %v", err)
		return "", "", "", ""
	}
	serverPublic := string(serverPublicKey)

	serverSecretKey, err := ioutil.ReadFile(privateKeysDir + "/server.key_secret")
	if err != nil {
		glog.Errorf("Error reading serverSecretKey : %v", err)
		return "", "", "", ""
	}
	serverSecret := string(serverSecretKey)

	return clientPublic, clientSecret, serverPublic, serverSecret
}

// Init method is used to initialize the socket
func Init(address string, publicKeysDir string, privateKeysDir string) (*zmq.Socket, error) {

	// Publisher for dev mode
	if publicKeysDir == "" && privateKeysDir == "" {

		//  Create socket for publish
		publisher, err := zmq.NewSocket(zmq.PUB)
		if err != nil {
			glog.Errorf("Socket creation failed : %v", err)
			return nil, err
		}
		//  Bind to the socket
		errSoc := publisher.Bind(address)
		if errSoc != nil {
			glog.Errorf("Error binding to socket : %v", errSoc)
			return nil, errSoc
		}

		return publisher, nil
	}

	// Load required certs
	clientPublic, _, _, serverSecret := loadCertificates(publicKeysDir, privateKeysDir)

	zmq.AuthSetVerbose(true)
	zmq.AuthStart()

	//  Tell authenticator to use this public client key
	zmq.AuthCurveAdd("domain1", clientPublic)

	//  Create socket for publish
	publisher, err := zmq.NewSocket(zmq.PUB)
	if err != nil {
		glog.Errorf("Socket creation failed : %v", err)
		return nil, err
	}

	publisher.ServerAuthCurve("domain1", serverSecret)

	//  Bind to the socket
	errSoc := publisher.Bind(address)
	if errSoc != nil {
		glog.Errorf("Error binding to socket : %v", errSoc)
		return nil, errSoc
	}

	return publisher, nil
}

// Send method is used to send the meta-data and image
func Send(publisher *zmq.Socket, listOfElements PublishList) error {

	glog.Infof("Publishing message %s on topic %s", listOfElements.MetaData, listOfElements.Topic)

	//  Send topic
	_, err := publisher.Send(listOfElements.Topic, zmq.SNDMORE)
	if err != nil {
		glog.Errorf("Error while sending topic : %v", err)
		return err
	}

	//  Send meta-data and frame
	_, errMsg := publisher.SendMessage(listOfElements.MetaData, listOfElements.FrameBlob)
	if err != nil {
		glog.Errorf("Error while sending JSON & buffer : %v", errMsg)
		return errMsg
	}

	time.Sleep(time.Second)
	return nil
}

// Close method is used to close the socket
func Close(publisher *zmq.Socket) {
	publisher.Close()
}

File: libs/ZmqLib/py/pub/publisher.py
import json
import os
import sys
import zmq
import zmq.auth
from zmq.auth.thread import ThreadAuthenticator


class ZMQ_PUBLISHER(object):
    """
    The ZMQ_PUBLISHER object to publish over a
    given topic
    """
    def __init__(self, address, public_keys_dir, secret_keys_dir):
        """
        Constructor to initialize the socket
        and create context
        """
        if public_keys_dir == "" and secret_keys_dir == "":
            self.dev_mode = True
        else:
            self.dev_mode = False

        if self.dev_mode:
            self.context = zmq.Context()
            self.publisher = self.context.socket(zmq.PUB)
            self.publisher.bind(address)
        else:
            # Check if certs exist
            if not (os.path.exists(public_keys_dir) and
                    os.path.exists(secret_keys_dir)):
                print("Certificates are missing - run generate_certificates.py\
                    script first")
                sys.exit(1)

            # Prepare our context and publisher
            self.context = zmq.Context()

            # Start an authenticator for this context.
            self.auth = ThreadAuthenticator(self.context)
            self.auth.start()

            # Tell authenticator to use the certificate in a directory
            self.auth.configure_curve(domain='*', location=public_keys_dir)
            self.publisher = self.context.socket(zmq.PUB)
            server_secret_file = os.path.join(secret_keys_dir,
                                              "zmq_server.key_secret")
            server_public, server_secret = \
                zmq.auth.load_certificate(server_secret_file)

            # Setting public and secret keys of server
            self.publisher.curve_secretkey = server_secret
            self.publisher.curve_publickey = server_public
            self.publisher.curve_server = True
            self.publisher.bind(address)

    def send(self, dataList=[], *args):
        """
        This method is used to publish the json meta-data
        and numpy frame

        Parameters :
        dataList : A list containing the topic, meta-data
                   and numpy frame to publish
        """
        flags = 0

        # Print the topic and message to be published
        print("Publish data %s over topic %s" % (dataList[1], dataList[0]))

        # Sending topic
        self.publisher.send_string(dataList[0], flags | zmq.SNDMORE)

        # Sending JSON meta-data
        self.publisher.send_json(dataList[1], flags | zmq.SNDMORE)

        # Sending numpy frame
        self.publisher.send(dataList[2], flags)

    def send_multipart(self, dataList=[], *args):
        """
        This method is used to publish data using send_multipart API

        Parameters :
        dataList : A list containing the elements to be published
        """

        # Print the topic and message to be published
        print("Publish data %s over topic %s" % (dataList[1], dataList[0]))

        # Sending topic
        self.publisher.send_multipart(dataList)

    def close_socket(self):
        # Close the socket and destroy the context
        self.publisher.close()
        self.context.term()
        if not self.dev_mode:
            self.auth.stop()
